### PR TITLE
Move `Gas` and `Motes` to casper-types.

### LINKED ISSUE #1841: Move Gas and Motes to casper_types

### PR DESCRIPTION
**Pull request by nilz3ro**
# About These Changes
* Moves `Gas` from EE Shared to casper-types.
* Moves `Motes` from EE Shared to casper-types.
* closes #1841

### CODE REVIEW

File: types/Cargo.toml [hunk 1]
```diff
@@ -17,10 +17,17 @@ bitflags = "1"
 blake2 = { version = "0.9.0", default-features = false }
 datasize = { version = "0.2.4", default-features = false }
 displaydoc = { version = "0.1", default-features = false, optional = true }
-ed25519-dalek = { version = "1.0.0", default-features = false, features = ["rand", "u64_backend"] }
+ed25519-dalek = { version = "1.0.0", default-features = false, features = [
+    "rand",
+    "u64_backend",
+] }
```
Comment (Fraser999) [inline types/Cargo.toml]: Let's not do this please unless the line becomes excessively long.  The only one which falls into that realm is the `std` feature collection.

File: types/src/gas.rs [hunk 2]
```diff
@@ -40,7 +48,7 @@ impl fmt::Display for Gas {
     }
 }
 
-impl std::ops::Add for Gas {
+impl core::ops::Add for Gas {
```
Comment (Fraser999) [inline types/src/gas.rs]: These should be imported so we only have `impl Add for Gas` please.

Same comment for the other `ops` types below.

File: types/src/motes.rs [hunk 3]
```diff
@@ -1,34 +1,40 @@
...
-use casper_types::{
+use crate::{
     bytesrepr::{self, FromBytes, ToBytes},
-    U512,
+    Gas, U512,
 };
 
-use crate::shared::gas::Gas;
-
+/// `Motes` are small amounts of cspr, expressed as `U512`.
```
Comment (Fraser999) [inline types/src/motes.rs]:
```suggestion
/// A struct representing a number of `Motes`.
```

File: types/src/gas.rs [hunk 4]
```diff
@@ -1,34 +1,42 @@
...
     pub fn new(value: U512) -> Self {
         Gas(value)
     }
 
+    /// Returns the inner `U512` value.
     pub fn value(&self) -> U512 {
         self.0
     }
 
+    /// Returns an `Option<Gas>`. Returns `None` if the conversion cannot be performed.
```
Comment (Fraser999) [inline types/src/gas.rs]:
```suggestion
    /// Converts the given `motes` to `Gas` by dividing them by `conv_rate`.
    ///
    /// Returns `None` if `conv_rate == 0`.
```

File: types/src/gas.rs [hunk 5]
```diff
@@ -1,34 +1,42 @@
...
     pub fn from_motes(motes: Motes, conv_rate: u64) -> Option<Self> {
         motes
             .value()
             .checked_div(U512::from(conv_rate))
             .map(Self::new)
     }
 
+    /// Safe addition method that returns an `Option<Gas>`. Returns `None` if an overflow has
+    /// occurred.
```
Comment (Fraser999) [inline types/src/gas.rs]: This is taken from the std lib docs:
```suggestion
    /// Checked integer addition. Computes `self + rhs`, returning `None` if overflow occurred.
```

File: types/src/gas.rs [hunk 6]
```diff
@@ -1,34 +1,42 @@
...
     pub fn checked_add(&self, rhs: Self) -> Option<Self> {
         self.0.checked_add(rhs.value()).map(Self::new)
     }
 
+    /// Safe subtraction method that returns an `Option<Gas>`. Returns `None` if an overflow has
+    /// occurred.
```
Comment (Fraser999) [inline types/src/gas.rs]:
```suggestion
    /// Checked integer subtraction. Computes `self - rhs`, returning `None` if overflow occurred.
```

File: types/src/motes.rs [hunk 7]
```diff
@@ -1,34 +1,40 @@
...
 )]
 pub struct Motes(U512);
 
 impl Motes {
+    /// Constructs a new `Motes`.
     pub fn new(value: U512) -> Motes {
         Motes(value)
     }
 
+    /// Safe addition method. Returns an `Option<Motes>`. Returns `None` if an overflow occurred.
```
Comment (Fraser999) [inline types/src/motes.rs]:
```suggestion
    /// Checked integer addition. Computes `self + rhs`, returning `None` if overflow occurred.
```

File: types/src/motes.rs [hunk 8]
```diff
@@ -1,34 +1,40 @@
...
     pub fn checked_add(&self, rhs: Self) -> Option<Self> {
         self.0.checked_add(rhs.value()).map(Self::new)
     }
 
+    /// Returns the `U512` value.
```
Comment (Fraser999) [inline types/src/motes.rs]:
```suggestion
    /// Returns the inner `U512` value.
```

File: types/src/motes.rs [hunk 9]
```diff
@@ -1,34 +1,40 @@
...
     pub fn value(&self) -> U512 {
         self.0
     }
 
+    /// Returns a `Option<Motes>`. Returns `None` if the conversion cannot be performed.
```
Comment (Fraser999) [inline types/src/motes.rs]:
```suggestion
    /// Converts the given `gas` to `Motes` by multiplying them by `conv_rate`.
    ///
    /// Returns `None` if an arithmetic overflow occurred.
```

File: types/src/motes.rs [hunk 10]
```diff
@@ -42,7 +48,7 @@ impl fmt::Display for Motes {
     }
 }
 
-impl std::ops::Add for Motes {
+impl core::ops::Add for Motes {
```
Comment (Fraser999) [inline types/src/motes.rs]: Let's import the `ops` types here too please.

**Review by Fraser999 — COMMENTED** [review]
Nitpicks only.

**Review by darthsiroftardis — APPROVED** [review]
Approved

**Pull request merged**